File: android/src/main/kotlin/com/gtbluesky/cooperation/Cooperation.kt
package com.gtbluesky.cooperation

import android.app.Application
import io.flutter.FlutterInjector
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.FlutterEngineGroup
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.plugin.common.MethodChannel

object Cooperation {
    private var engine: FlutterEngine? = null
    private var methodChannel: MethodChannel? = null

    fun install(engine: FlutterEngine) {
        this.engine = engine
        methodChannel = MethodChannel(engine.dartExecutor, "cooperation_channel")
    }

    fun install(
        context: Application,
        engineGroup: FlutterEngineGroup,
        library: String? = null,
        entryPoint: String = "main"
    ) {
        val flutterLoader = FlutterInjector.instance().flutterLoader()
        val dartEntrypoint = if (library == null) {
            DartExecutor.DartEntrypoint(flutterLoader.findAppBundlePath(), entryPoint)
        } else {
            DartExecutor.DartEntrypoint(flutterLoader.findAppBundlePath(), library, entryPoint)
        }
        val engine = engineGroup.createAndRunEngine(context, dartEntrypoint)
        install(engine)
//        GeneratedPluginRegister.registerGeneratedPlugins(engine)
    }

    fun uninstall() {
        methodChannel = null
        engine?.destroy()
        engine = null
    }

    fun call(
        method: String,
        arguments: Map<String, Any>? = null,
        result: MethodChannel.Result? = null
    ) {
        methodChannel?.invokeMethod(method, arguments, result)
    }
}

File: ios/Classes/Cooperation.swift
//
// Created by gtbluesky on 2022/10/9.
//

import Foundation

public class Cooperation: NSObject {
    @objc public static let instance = Cooperation()
    private var engine: FlutterEngine? = nil
    private var methodChannel: FlutterMethodChannel? = nil

    private override init() {
        super.init()
    }

    @objc public func install(_ engine: FlutterEngine) {
        self.engine = engine
        methodChannel = FlutterMethodChannel(name: "cooperation_channel", binaryMessenger: engine.binaryMessenger)
    }

    @objc public func install(
            engineGroup: FlutterEngineGroup,
            library: String? = nil,
            entryPoint: String? = nil
    ) {
        let engine = engineGroup.makeEngine(withEntrypoint: entryPoint, libraryURI: library)
        install(engine)
        let clazz = NSClassFromString("GeneratedPluginRegistrant") as? NSObject.Type
        let selector = NSSelectorFromString("registerWithRegistry:")
        if clazz?.responds(to: selector) == true {
            clazz?.perform(selector, with: engine)
        }
    }

    @objc public func uninstall() {
        methodChannel = nil
        engine?.destroyContext()
        engine = nil
    }

    @objc public func call(
            method: String,
            arguments: Dictionary<String, Any>? = nil,
            result: FlutterResult? = nil
    ) {
        methodChannel?.invokeMethod(method, arguments: arguments, result: result)
    }
}